keps/sig-architecture/3784-database-of-names/README.md

```diff
@@ -0,0 +1,449 @@
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Kubernetes has many extensible fields which can carry either "standard"
+(contextual) or user-defined values.  Examples include (but are not limited to)
+label keys, annotation keys, node taints, pod resource names, conditions, etc.
+
+In most cases we define a "reserved" block, which is set aside for the project
+to use.  Labels and annotations which are prefixed with a `k8s.io` or
+`kubernetes.io` domain are reserved, and should not be used by end-users,
+whereas unprefixed names are reserved for end-users and should not be used by
+Kubernetes or any other project.
+
+We have made great use of these reserved names in myriad ways, all over the
+project.  What we have not had is a common way to know WHICH names we are
+using, what they mean, where they apply, or who is responsible for them.  In
+fact, the only way to find them is to read human-oriented documentation or to
+grep the codebase(s).
+
+This KEP proposes to store these names in a way that is searchable, machine
+parseable, but also relatively human-friendly.
+
+## Motivation
+
+We have too many KEPs and PRs that add annotation- or label-keys without any
+sort of collision detection or governance or documentation.
+
+### Goals
+
+1. Publish a list of known values.
+1. Have a place where end-users can go to find information.
+1. Have a place we can send developers to get some modicum of review of these
+names.
+
+#### Maybe later
+
+1. Add this to the KEP template
+1. Extend this DB to highly-coupled projects (e.g. containerd, crio)
+1. Extend this DB to loosely-coupled projects (e.g. GKE, EKS, AKS)
+1. Use in-code metadata to auto-generate these DB records
+
+### Non-Goals
+
+1. Completeness.  We know that getting 100% of the worms back into the can may
+   never happen.  That doesn't mean we should not try.
+1. Enforcement.  We do not intend to inject this into code-review processes.
+   We do intend to make people **aware** of it, so they can voluntarily contribute.
+
+## Proposal
+
+The "database" should live in a new git repo (proposed name:
+"kubernetes/namesdb").  In this repo we will establish a files-and-directories
```

Comment:
We have reserved things besides names too. eg https://github.com/kubernetes/kubernetes/blob/master/pkg/cluster/ports/ports.go. Would it be useful to step back one level and just have a generic "kubernetes reserved things" repository ("Kubernetes Assigned Numbers Authority"?), with names being one subdir and others potentially added later?

Reply:
I don't know that ports need all the same tooling and metadata, but I agree we don't do a great job of documenting them.  Would it be enough to just convert that .go file into markdown for k/website?

Reply:
> Would it be enough to just convert that .go file into markdown for k/website?

We (SIG Docs) can also work with JSON or YAML. JSON is handy because it has a defined media type and we can do HTTP fetches.

Reply:
I was not concerned _specifically_ with ports being documented/parseable, I was just wondering if we're going to need a bigger registry later, and thinking it would be nicer to have it all in one repo rather than having both a "namesdb" and a "portsdb" and who knows what else (Kubernetes-specific IPFIX elements?)

(Though the fact that ports was the first thing that came to mind is because [OpenShift _does_ have a centralized port registry](https://github.com/openshift/enhancements/blob/master/dev-guide/host-port-registry.md) because random operators kept trying to claim the same metrics ports. But this is a bigger management problem in OpenShift because it's spread across so many repos.)

Reply:
ACK - I don't object to leaving room at the root for an alternate hierarchy of of data (e.g.  .../names/... vs .../ports/...) with different schema.

Reply:
That changes "namesdb" to ... "iddb" or "nndb" (names and numbers) or "metadb" or something?